services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+      }
+    }
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    executor.scheduleAtFixedRate(
+        new ProtocolVersionDetectionTask(),
+        0,
+        multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionIntervalMs(),
+        TimeUnit.MILLISECONDS);
+    LOGGER.info("ProtocolVersionAutoDetectionService is started");
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop.set(true);
+    executor.shutdown();
+
+    try {
+      // Wait for the executor to terminate
+      if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+        executor.shutdownNow(); // Force shutdown if not terminated in 60 seconds
+      }
+    } catch (InterruptedException e) {
+      LOGGER.error("Error shutting down ProtocolVersionAutoDetectionService", e);
+      executor.shutdownNow();
+    }
+  }
+
+  private ProtocolVersionAutoDetectionStats getClusterStat(String clusterName) {
+    return clusterToStatsMap.get(clusterName);
+  }
+
+  class ProtocolVersionDetectionTask implements Runnable {
+    @Override
+    public void run() {
+      LOGGER.info("Started running {}", getClass().getSimpleName());
+      while (!stop.get()) {
+        for (String clusterName: allClusters) {
+          if (!veniceParentHelixAdmin.isLeaderControllerFor(clusterName)) {
+            continue;
+          }
+          try {
+            // start the clock
+            long startTime = System.currentTimeMillis();
+            Long currentGoodVersion = getSmallestLocalAdminOperationProtocolVersionForAllConsumers(clusterName);
+            Long upstreamVersion = AdminTopicMetadataAccessor.getAdminOperationProtocolVersion(
+                veniceParentHelixAdmin.getAdminTopicMetadata(clusterName, Optional.empty()));
+            LOGGER.info(
+                "Current good Admin Operation version for cluster {} is {} and upstream version is {}",
+                clusterName,
+                currentGoodVersion,
+                upstreamVersion);
+            if (upstreamVersion != -1 && !Objects.equals(currentGoodVersion, upstreamVersion)) {
+              veniceParentHelixAdmin.updateAdminOperationProtocolVersion(clusterName, currentGoodVersion);
+              LOGGER.info(
+                  "Updated admin operation protocol version in ZK for cluster {} from {} to {}",
+                  clusterName,
+                  upstreamVersion,
+                  currentGoodVersion);
+            }
+            long elapsedTimeFromMsToMs = LatencyUtils.getElapsedTimeFromMsToMs(startTime);
+            getClusterStat(clusterName).recordProtocolVersionAutoDetectionLatencySensor(elapsedTimeFromMsToMs);
+          } catch (Exception e) {
+            LOGGER.warn("Received exception while running ProtocolVersionDetectionTask", e);
+            getClusterStat(clusterName).recordProtocolVersionAutoDetectionErrorSensor();
+          } catch (Throwable throwable) {
```

Comment:
I see. If we capturing throwable anyway, you can remove the Exception catch block.
Would InterruptException be thrown from this code block? If so, please don't swallow InterruptException.

Reply:
I keep Exception and remove Throwable. Error that is Throwable and not Exception is very rare, catching Exception should be enough.